### PR TITLE
add equality for ingressrule

### LINKED ISSUE #145: Implement equality method for Ingress Rule
/kind feature

Follow up from https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/141#discussion_r222014810

### PR DESCRIPTION
**Pull request by vjdhama**
**What this PR does / why we need it**:

Implements equality for IngressRule struct.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #145

**Release note**:
None

cc/ @vincepri @chuckha

### CODE REVIEW

File: pkg/apis/awsprovider/v1alpha1/types.go [hunk 1]
```diff
@@ -340,6 +335,40 @@ func (i IngressRules) Difference(o IngressRules) (out IngressRules) {
 	return
 }
 
+// Equal returns true if two IngressRule are equal
+func (i *IngressRule) Equal(o *IngressRule) bool {
```
Comment (ashish-amarnath) [inline pkg/apis/awsprovider/v1alpha1/types.go]: Can you please add some unit tests for this?

Comment (vjdhama) [reply pkg/apis/awsprovider/v1alpha1/types.go]: Sure, I'll add.

Comment (vjdhama) [reply pkg/apis/awsprovider/v1alpha1/types.go]: @ashish-amarnath I have added tests. PTAL.

Comment (ashish-amarnath) [reply pkg/apis/awsprovider/v1alpha1/types.go]: @vjdhama  Thank you for adding the unit tests :-)

File: pkg/apis/awsprovider/v1alpha1/types_test.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+				Protocol:               providerv1.SecurityGroupProtocolTCP,
+				FromPort:               0,
+				ToPort:                 80,
+				CidrBlocks:             []string{"172.16.12.0/24", "10.20.30.0/24", "10.30.40.0/24"},
+				SourceSecurityGroupIDs: []string{"sg-1234567", "sg-1236745"},
+			},
+			expected: true,
+		},
+		{
+			name: "Equal should return true for same SourceSecurityGroupIDs",
```
Comment (ashish-amarnath) [inline pkg/apis/awsprovider/v1alpha1/types_test.go]: How is this test different from the previous one?

Also, looks like you'll have 8 cases returning false and only one case returning true.
For return false case:
* Different `SourceSecurityGroupIDs ` and `CidrBlocks` lengths case aren't covered.
* Different `SourceSecurityGroupIDs ` and `CidrBlocks` of equal length aren't covered.
Can you please add those?

File: pkg/apis/awsprovider/v1alpha1/types_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+			},
+			expected: true,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := tc.inputIngressRule.Equal(tc.inputIngressRuleParam)
+			if !reflect.DeepEqual(actual, tc.expected) {
+				t.Fatalf("expected equality but got\n%#v\nand expected\n%#v", actual, tc.expected)
```
Comment (ashish-amarnath) [inline pkg/apis/awsprovider/v1alpha1/types_test.go]: Please follow this guide for the unit test failure messages
https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

You can maybe say `t.Fatalf("For [%s], equality check failed, Got: %t, want: %t", tc.name, actual, tc.expected)`

File: pkg/apis/awsprovider/v1alpha1/types_test.go (hunk 2 above)
Comment (vjdhama) [reply pkg/apis/awsprovider/v1alpha1/types_test.go]: Shoot. I'll add those cases as well.

File: pkg/apis/awsprovider/v1alpha1/types_test.go (hunk 3 above)
Comment (vjdhama) [reply pkg/apis/awsprovider/v1alpha1/types_test.go]: Sure, will do that. Thanks.

File: pkg/apis/awsprovider/v1alpha1/types.go (hunk 1 above)
Comment (vincepri) [inline pkg/apis/awsprovider/v1alpha1/types.go]: This method should probably be named `Equals`, wdyt?

Comment (vjdhama) [reply pkg/apis/awsprovider/v1alpha1/types.go]: Now, that I think of it, you're right. Let me change to `Equals`.

**Comment by vjdhama** [comment]
@ashish-amarnath @vincepri Made the necessary changes. PTAL.

**Pull request merged**